source/common/router/router.cc

```diff
@@ -548,8 +548,11 @@ Http::ConnectionPool::Instance* Filter::getConnPool() {
     const auto& alpn =
         callbacks_->streamInfo().filterState().getDataReadOnly<Network::ApplicationProtocols>(
             Network::ApplicationProtocols::key());
-    transport_socket_options_ = std::make_shared<Network::TransportSocketOptionsImpl>(
-        "", std::vector<std::string>{}, std::vector<std::string>{alpn.value()});
+
+    if (alpn.hasProtocol(protocol)) {
```

Comment:
FWIW I'm not thrilled that logic this specific is creeping into the router filter. Can we instead handle this in a similar fashion to: 

https://github.com/envoyproxy/envoy/blob/4e858f17fe08224c9c089240908ccd0c518e01a7/include/envoy/http/filter.h#L426-L438

But instead for transport socket options? I think this should be a very similar situation?

cc @lizan can you take a look at ^ and guide in that direction depending on what you think? Thank you!

Reply:
Yes I think the logic should be in extension (filter) code, then the transport socket options should be modified through filter callbacks.

Reply:
The filter has limited knowledge to set the transport socket option. The filter only knows the downstream protocol and it can only select the ALPN according to the downstream protocol but I think the ALPN should be selected according to the upstream protocol instead of downstream's. e.g., if the downstream connection uses HTTP/1.1 and upstream connection uses HTTP/2, the ALPN override will be wrong.

Reply:
The filter will know the upstream cluster that the the request is going to be routed to. This should be sufficient to decide at that point which ALPN to use, no? I think this can all be moved into a custom filter other than allowing filters to modify the transport socket options.

Reply:
+1 for @yxue 
Since router decide if the upstream http protocol is h1 or h2,
ALPN http filter should not set the final alpn, it should provide two pool: h1 and h2

Reply:
The router does not decide whether upstream is h1 or h2, the cluster and cluster manager does. I'm not sure I see what benefit having this logic in the router provides. By mimicking how socket options are set by the filter, the cluster manager connection pool implemention can then do the right thing.

Reply:
@mattklein123 router does based on cluster: https://github.com/envoyproxy/envoy/blob/master/source/common/router/router.cc#L568 because we have `USE_DOWNSTREAM_PROTOCOL` and router is the place to inject that.

Reply:
Hmm ok. I forgot about this feature. Alright this all strikes me as suboptimal but I'm ok with leaving it for now. Can you add a TODO that we might want to rethink where this logic lives in the future? Also, I would still prefer that we add generic facilities for filters to modify transport socket options and then just use that from the router for now, but I will defer to Lizan on whether to do that now or put in a TODO.

Reply:
So here is the PR to let filter access upstream protocol https://github.com/envoyproxy/envoy/pull/8544, @yxue with this I believe you don't need to put the logic in router anymore.

> Also, I would still prefer that we add generic facilities for filters to modify transport socket options and then just use that from the router for now.

I'll do this in context of #2690.

Reply:
IIUC, we could do that without putting the upstream protocol in the cluster info because the http filter can also access the cluster and route info.  Is the reason for the PR that we want to have single source of truth?

Reply:
Yes, there’s another PR to add Http3 and I expect that logic can be more complicated in the future.

Reply:
> Also, I would still prefer that we add generic facilities for filters to modify transport socket options and then just use that from the router for now.

Unlike socket options its not very generic so I think keep using filter state is fine, so:
https://github.com/envoyproxy/envoy/pull/8546

Reply:
with #8546 and #8544, an http filter can calculate alpn override based on either upstream or downstream protocol and set it to filter state, can we close this? @yxue 

I'm inclining to have that filter in istio/proxy because it seems very istio specific. If we see other interest then we can move.

Reply:
@lizan thanks for PRs. I will close this one and another HTTP filter pr. I already created the PR in istio/proxy https://github.com/istio/proxy/pull/2450